// jasmine-spec-reporter/built/configuration-parser.js
"use strict";
var ConfigurationParser = (function () {
    function ConfigurationParser() {
    }
    ConfigurationParser.parse = function (conf) {
        return ConfigurationParser.merge(ConfigurationParser.defaultConfiguration, conf);
    };
    ConfigurationParser.merge = function (template, override) {
        var result = {};
        for (var key in template) {
            if (template[key] instanceof Object
                && !(template[key] instanceof Array)
                && override instanceof Object
                && override[key] instanceof Object
                && !(override[key] instanceof Array)) {
                result[key] = ConfigurationParser.merge(template[key], override[key]);
            }
            else if (override instanceof Object
                && Object.keys(override).indexOf(key) !== -1) {
                result[key] = override[key];
            }
            else {
                result[key] = template[key];
            }
        }
        if (override instanceof Object && override.customOptions) {
            result.customOptions = override.customOptions;
        }
        return result;
    };
    ;
    return ConfigurationParser;
}());
ConfigurationParser.isWindows = process && process.platform === "win32";
ConfigurationParser.defaultConfiguration = {
    suite: {
        displayNumber: false,
    },
    spec: {
        displayStacktrace: false,
        displaySuccessful: true,
        displayFailed: true,
        displayPending: false,
        displayDuration: false,
    },
    summary: {
        displayStacktrace: false,
        displaySuccessful: false,
        displayFailed: true,
        displayPending: true,
    },
    colors: {
        enabled: true,
        successful: "green",
        failed: "red",
        pending: "yellow"
    },
    prefixes: {
        successful: ConfigurationParser.isWindows ? "\u221A " : "✓ ",
        failed: ConfigurationParser.isWindows ? "\u00D7 " : "✗ ",
        pending: "* "
    },
    customProcessors: []
};
exports.ConfigurationParser = ConfigurationParser;
