// jasmine-spec-reporter/built/execution-display.js
"use strict";
var default_processor_1 = require("./display/processors/default-processor");
var spec_colors_processor_1 = require("./display/processors/spec-colors-processor");
var spec_durations_processor_1 = require("./display/processors/spec-durations-processor");
var spec_prefixes_processor_1 = require("./display/processors/spec-prefixes-processor");
var suite_numbering_processor_1 = require("./display/processors/suite-numbering-processor");
var colors_display_1 = require("./display/colors-display");
var ExecutionDisplay = (function () {
    function ExecutionDisplay(configuration) {
        this.configuration = configuration;
        this.indent = "  ";
        this.currentIndent = "";
        this.suiteHierarchy = [];
        this.suiteHierarchyDisplayed = [];
        this.successfulSpecs = [];
        this.failedSpecs = [];
        this.pendingSpecs = [];
        this.lastWasNewLine = false;
        colors_display_1.ColorsDisplay.init(this.configuration);
        this.displayProcessors = ExecutionDisplay.initProcessors(this.configuration);
        this.hasCustomDisplaySpecStarted = ExecutionDisplay.hasCustomDisplaySpecStarted(this.displayProcessors);
    }
    ExecutionDisplay.initProcessors = function (configuration) {
        var displayProcessors = [
            new default_processor_1.DefaultProcessor(configuration),
            new spec_prefixes_processor_1.SpecPrefixesProcessor(configuration),
            new spec_colors_processor_1.SpecColorsProcessor(configuration)
        ];
        if (configuration.spec.displayDuration) {
            displayProcessors.push(new spec_durations_processor_1.SpecDurationsProcessor(configuration));
        }
        if (configuration.suite.displayNumber) {
            displayProcessors.push(new suite_numbering_processor_1.SuiteNumberingProcessor(configuration));
        }
        if (configuration.customProcessors) {
            configuration.customProcessors.forEach(function (Processor) {
                displayProcessors.push(new Processor(configuration));
            });
        }
        return displayProcessors;
    };
    ExecutionDisplay.hasCustomDisplaySpecStarted = function (processors) {
        var isDisplayed = false;
        processors.forEach(function (processor) {
            var log = "foo";
            var result = processor.displaySpecStarted({ id: "bar", description: "bar", fullName: "bar" }, log);
            isDisplayed = isDisplayed || result !== log;
        });
        return isDisplayed;
    };
    ExecutionDisplay.prototype.jasmineStarted = function (runner) {
        var log = "";
        this.displayProcessors.forEach(function (displayProcessor) {
            log = displayProcessor.displayJasmineStarted(runner, log);
        });
        this.log(log);
    };
    ExecutionDisplay.prototype.summary = function (metrics) {
        var execution = "Executed " + metrics.executedSpecs + " of " + metrics.totalSpecsDefined + (metrics.totalSpecsDefined === 1 ? " spec " : " specs ");
        var successful = (metrics.failedSpecs === 0) ? "SUCCESS " : "";
        var failed = (metrics.failedSpecs > 0) ? "(" + metrics.failedSpecs + " FAILED) " : "";
        var pending = (metrics.pendingSpecs > 0) ? "(" + metrics.pendingSpecs + " PENDING) " : "";
        var skipped = (metrics.skippedSpecs > 0) ? "(" + metrics.skippedSpecs + " SKIPPED) " : "";
        var duration = "in " + metrics.duration + ".";
        this.resetIndent();
        this.newLine();
        if (this.configuration.summary.displaySuccessful && metrics.successfulSpecs > 0) {
            this.successesSummary();
        }
        if (this.configuration.summary.displayFailed && metrics.failedSpecs > 0) {
            this.failuresSummary();
        }
        if (this.configuration.summary.displayPending && metrics.pendingSpecs > 0) {
            this.pendingsSummary();
        }
        this.log(execution + successful.successful + failed.failed + pending.pending + skipped + duration);
        if (metrics.random) {
            this.log("Randomized with seed " + metrics.seed + ".");
        }
    };
    ExecutionDisplay.prototype.successesSummary = function () {
        this.log("**************************************************");
        this.log("*                   Successes                    *");
        this.log("**************************************************");
        this.newLine();
        for (var i = 0; i < this.successfulSpecs.length; i++) {
            this.successfulSummary(this.successfulSpecs[i], i + 1);
            this.newLine();
        }
        this.newLine();
        this.resetIndent();
    };
    ExecutionDisplay.prototype.successfulSummary = function (spec, index) {
        this.log(index + ") " + spec.fullName);
    };
    ExecutionDisplay.prototype.failuresSummary = function () {
        this.log("**************************************************");
        this.log("*                    Failures                    *");
        this.log("**************************************************");
        this.newLine();
        for (var i = 0; i < this.failedSpecs.length; i++) {
            this.failedSummary(this.failedSpecs[i], i + 1);
            this.newLine();
        }
        this.newLine();
        this.resetIndent();
    };
    ExecutionDisplay.prototype.failedSummary = function (spec, index) {
        this.log(index + ") " + spec.fullName);
        this.displayErrorMessages(spec, this.configuration.summary.displayStacktrace);
    };
    ExecutionDisplay.prototype.pendingsSummary = function () {
        this.log("**************************************************");
        this.log("*                    Pending                     *");
        this.log("**************************************************");
        this.newLine();
        for (var i = 0; i < this.pendingSpecs.length; i++) {
            this.pendingSummary(this.pendingSpecs[i], i + 1);
            this.newLine();
        }
        this.newLine();
        this.resetIndent();
    };
    ExecutionDisplay.prototype.pendingSummary = function (spec, index) {
        this.log(index + ") " + spec.fullName);
        this.increaseIndent();
        var pendingReason = spec.pendingReason ? spec.pendingReason : "No reason given";
        this.log(pendingReason.pending);
        this.resetIndent();
    };
    ExecutionDisplay.prototype.specStarted = function (spec) {
        if (this.hasCustomDisplaySpecStarted) {
            this.ensureSuiteDisplayed();
            var log_1 = "";
            this.displayProcessors.forEach(function (displayProcessor) {
                log_1 = displayProcessor.displaySpecStarted(spec, log_1);
            });
            this.log(log_1);
        }
    };
    ExecutionDisplay.prototype.successful = function (spec) {
        this.successfulSpecs.push(spec);
        if (this.configuration.spec.displaySuccessful) {
            this.ensureSuiteDisplayed();
            var log_2 = "";
            this.displayProcessors.forEach(function (displayProcessor) {
                log_2 = displayProcessor.displaySuccessfulSpec(spec, log_2);
            });
            this.log(log_2);
        }
    };
    ExecutionDisplay.prototype.failed = function (spec) {
        this.failedSpecs.push(spec);
        if (this.configuration.spec.displayFailed) {
            this.ensureSuiteDisplayed();
            var log_3 = "";
            this.displayProcessors.forEach(function (displayProcessor) {
                log_3 = displayProcessor.displayFailedSpec(spec, log_3);
            });
            this.log(log_3);
            this.displayErrorMessages(spec, this.configuration.spec.displayStacktrace);
        }
    };
    ExecutionDisplay.prototype.pending = function (spec) {
        this.pendingSpecs.push(spec);
        if (this.configuration.spec.displayPending) {
            this.ensureSuiteDisplayed();
            var log_4 = "";
            this.displayProcessors.forEach(function (displayProcessor) {
                log_4 = displayProcessor.displayPendingSpec(spec, log_4);
            });
            this.log(log_4);
        }
    };
    ExecutionDisplay.prototype.displayErrorMessages = function (spec, withStacktrace) {
        this.increaseIndent();
        for (var i = 0; i < spec.failedExpectations.length; i++) {
            this.log("- ".failed + spec.failedExpectations[i].message.failed);
            if (withStacktrace && spec.failedExpectations[i].stack) {
                this.log(this.filterStackTraces(spec.failedExpectations[i].stack));
            }
        }
        this.decreaseIndent();
    };
    ExecutionDisplay.prototype.filterStackTraces = function (traces) {
        var lines = traces.split("\n");
        var filtered = [];
        for (var i = 1; i < lines.length; i++) {
            if (!/(jasmine[^\/]*\.js|Timer\.listOnTimeout)/.test(lines[i])) {
                filtered.push(lines[i]);
            }
        }
        return filtered.join("\n" + this.currentIndent);
    };
    ExecutionDisplay.prototype.suiteStarted = function (suite) {
        this.suiteHierarchy.push(suite);
    };
    ExecutionDisplay.prototype.suiteDone = function () {
        var suite = this.suiteHierarchy.pop();
        if (this.suiteHierarchyDisplayed[this.suiteHierarchyDisplayed.length - 1] === suite) {
            this.suiteHierarchyDisplayed.pop();
        }
        this.newLine();
        this.decreaseIndent();
    };
    ExecutionDisplay.prototype.ensureSuiteDisplayed = function () {
        if (this.suiteHierarchy.length !== 0) {
            for (var i = this.suiteHierarchyDisplayed.length; i < this.suiteHierarchy.length; i++) {
                this.suiteHierarchyDisplayed.push(this.suiteHierarchy[i]);
                this.displaySuite(this.suiteHierarchy[i]);
            }
        }
        else {
            var topLevelSuite = { description: "Top level suite" };
            this.suiteHierarchy.push(topLevelSuite);
            this.suiteHierarchyDisplayed.push(topLevelSuite);
            this.displaySuite(topLevelSuite);
        }
    };
    ExecutionDisplay.prototype.displaySuite = function (suite) {
        this.newLine();
        this.computeSuiteIndent();
        var log = "";
        this.displayProcessors.forEach(function (displayProcessor) {
            log = displayProcessor.displaySuite(suite, log);
        });
        this.log(log);
        this.increaseIndent();
    };
    ExecutionDisplay.prototype.computeSuiteIndent = function () {
        this.resetIndent();
        for (var i = 0; i < this.suiteHierarchyDisplayed.length; i++) {
            this.increaseIndent();
        }
    };
    ExecutionDisplay.prototype.log = function (stuff) {
        if (stuff !== null) {
            console.log(this.currentIndent + stuff);
            this.lastWasNewLine = false;
        }
    };
    ExecutionDisplay.prototype.newLine = function () {
        if (!this.lastWasNewLine) {
            console.log("");
            this.lastWasNewLine = true;
        }
    };
    ExecutionDisplay.prototype.resetIndent = function () {
        this.currentIndent = "";
    };
    ExecutionDisplay.prototype.increaseIndent = function () {
        this.currentIndent += this.indent;
    };
    ExecutionDisplay.prototype.decreaseIndent = function () {
        this.currentIndent = this.currentIndent.substr(0, this.currentIndent.length - this.indent.length);
    };
    return ExecutionDisplay;
}());
exports.ExecutionDisplay = ExecutionDisplay;
